File: findSecondMinimumValue.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

//19.63
class Solution {
    public int findSecondMinimumValue(TreeNode root) {
        if(root == null)
            return -1;
        Queue<TreeNode> que = new LinkedList<TreeNode>();
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
        que.add(root);
        //TreeNode tmp;
        for(TreeNode tmp = que.peek(); !que.isEmpty(); tmp = que.peek()){
            if(!pq.contains(tmp.val))
                pq.add(tmp.val);
            if(tmp.left != null)
                que.add(tmp.left);
            if(tmp.right != null)
                que.add(tmp.right);
            que.poll();
        }
        
        if(pq.size() < 2)
            return -1;
        
        pq.poll();
        return pq.peek();
    }
}

//8.01
class Solution {
    public int findSecondMinimumValue(TreeNode root) {
        if(root == null)
            return -1;
        Queue<TreeNode> que = new LinkedList<TreeNode>();
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        que.add(root);
        //TreeNode tmp;
        for(TreeNode tmp = que.peek(); !que.isEmpty(); tmp = que.peek()){
            if(map.get(tmp.val) == null){
                pq.add(tmp.val);
                map.put(tmp.val, 1);
            }
            if(tmp.left != null)
                que.add(tmp.left);
            if(tmp.right != null)
                que.add(tmp.right);
            que.poll();
        }
        
        if(pq.size() < 2)
            return -1;
        
        pq.poll();
        return pq.peek();
    }
}

File: findSecondMinimumValue.cpp
#include<cstdlib>
#include<iostream>
#include<vector>
#include<cstring>
#include<algorithm>
#include<string>
#include<sstream>
#include<math.h>
#include<climits>
#include <sstream>


using namespace std;
const int size =9;

//100
class Solution {
public:
    int findSecondMinimumValue(TreeNode* root) {
        unordered_map<int, int> map;
        priority_queue<int, vector<int>, greater<int>> pq;
        
        queue<TreeNode*> que;
        if(root != nullptr)
            que.push(root);
        for(;!que.empty(); que.pop()){
            TreeNode* tmpTree = que.front();
            if(map[tmpTree->val]++ == 0)
                pq.push(tmpTree->val);
            if(tmpTree->left != nullptr)
                que.push(tmpTree->left);
            if(tmpTree->right != nullptr)
                que.push(tmpTree->right);
        }
        if(pq.size() < 2)
            return -1;
        
        pq.pop();
        return pq.top();
    }
};

int main(){
	vector<int> nums = {1,1,1,0};
	int target = -100, ans;
	ans = threeSumClosest(nums, target);
	cout<<"ans = "<<ans<<endl;
	return EXIT_SUCCESS;
}
